강원대 BE_최현준 Step3 - 위시리스트 구현

Hyeonjun0527 opened the pull request:
## Step 3: 위시리스트 기능 개발

### 주요 변경 사항

1. **위시리스트(Wish) 도메인**
    - `Wish` 도메인 생성
    - 어플리케이션 계층의 비즈니스 로직 : `WishUseCase` 인터페이스(Port) 그리고 구현체인 `WishInteractor`를 추가.
    - 데이터 영속성을 처리하기 위해 `WishPersistencePort`와 JPA 기반의 세컨더리 어댑터(`WishPersistenceAdapter`, `WishJpaRepository`)를 구현.

2. **위시리스트 API 엔드포인트 구현**
    - `WishController` (REst)
        - `GET /api/wishes`: 현재 로그인한 사용자의 위시리스트 목록을 페이지네이션으로 조회
        - `POST /api/wishes`: 특정 상품을 위시리스트에 추가
        - `PUT /api/wishes/{id}/quantity`: 위시리스트에 담긴 상품의 수량을 변경
        - `DELETE /api/wishes/{id}`: 특정 위시리스트 항목을 삭제
    - `@LoginMember` 어노테이션을 활용하여 각 요청에 대해 JWT 토큰에서 사용자 정보를 자동으로 주입

3. **프론트엔드 UI/UX 구현 (`index.html`)**
    - 메인 페이지에 '내 위시리스트'  추가

4. **클라이언트-서버 연동 로직 구현 (`app.js`)**
    - 페이지 로드 시 위시리스트 데이터를 비동기적으로 가져오는 `fetchWishes` 함수 구현
    - 가져온 데이터를 HTML 요소로 변환하여 화면에 표시하는 `renderWishes` 함수 구현
    - 상품 목록의 '위시리스트에 추가' 버튼과 연동되는 `addWish` 함수, 수량 변경을 위한 `updateWishQuantity` 함수, 삭제를 위한 `deleteWish` 함수

---

### 맞닥뜨린 문제 및 한계

1. **반복적인 NaN 페이지네이션 오류 및 해결**
`app.js`의 상품 목록과 `admin.js`의 관리자 목록에서 발생했던 `NaN` 관련 오류가 계속 발생. 
 페이지네이션 로직이 초기 페이지(page number가 0 또는 undefined)를 처리하지 못함.
pageData.number가 undefined 또는 null이여서 여기에 +1 -1 을 해서 NaN이 되는 것이였음.
관리자페이지에서도 똑같은 문제 발생.
중복되는 문제를 `common.js`로 공통으로 관리함

const currentPage = pageData.number || 0;
→ pageData.number가 undefined/null이면 0으로 대체

2. 관리자 페이지 버튼을 눌러도 이동되지 않는 문제

기존에는 skip path에 /admin을 두지 않았음.하지만 그렇게 하니 관리자도 접속이 불가했음
브라우저의 기본동작은 로컬스토리지에 저장된 값을 서버로 보내지 않아 새로고침시 
인증과정을 정상적으로 진행할 수 없었음. 

그래서 skip path 에 /admin을 두되,
관리자페이지로드되면 즉시 admin.js스크립트
실행하여 로컬스토리지 검사하여 /login으로 리다이렉트하도록 리팩토링함.
그리고 쿠키방식이 필요하다고 느낌.

---

### 중점적으로 리뷰 받고 싶은 곳

관리자에서 /admin 페이지를 SKIP_PATH로 허용하고 로컬스토리지에 값을 꺼내 검증하고 검증이 안되면 리다이렉트를 하도록 했는데 보안 문제가 발생하지 않을지.

아예 /admin 페이지를 404페이지 반환으로 막으니 관리자도 /admin페이지를 브라우저에 직접 입력하거나
새로고침을 할 시 토큰 없는 요청이 서버로 가게되어 404를 반환하게 됩니다.
그래서 쿠키나 세션이 반드시 필요하다고 느끼게 되었습니다.
쿠키는 무조건 브라우저가 자동으로 전송하니
그래서 자바스크립트 공격에 안전하게 서비스를 구축하려면
로컬 스토리지가 아닌 jwt를 Http-only Cookie 방식으로 브라우저에 저장을 해야되겠다고 느꼈습니다. 

현재 jwt필터와 인터셉터가 제 역할을 하고 있는지

wish 도메인에서 리팩토링할 만 한 요소

---

### 화면 테스트 방법
data.sql에서 관리자 아이디 확인 가능
admin@admin.com
1234
로 관리자 계정으로 접속 가능. 
/ : 상품추가 및 위시리스트 추가, 조회가 가능한 메인페이지
/admin : 상품 CRUD가 가능하고 회원리스트 볼 수 있는 관리자 페이지
/login : 로그인페이지

### 2차 리뷰

모든 리뷰 반영

- UnauthorizedException 로 에러메세지 변경
- WishPersistenceAdapter 내부에 RowMapper 리팩토링
  - WISH_WITH_DETAILS_ROW_MAPPER 추가 중복 로직 공통화
- JwtAuthenticationFilter의
  - sendUnauthorizedResponse → sendUnAuthenticationResponse로 이름 변경
  - PUBLIC_PATTERNS로 public endpoint 패턴 등록 및 AntPathMatcher로 필터 처리 일관성 강화
- @min, https://github.com/NotNull 어노테이션 추가
- Member/Adapter/Persistence 경로 구조 리팩토링
- 잘못만든 어댑터 삭제후 persistenceAdapter → adapter로 경로 변경

Hyeonjun0527 committed:
refactor(product): 유스케이스 계층 도입 및 아키텍처 개선

Product 관련 기능 유지보수성, 테스트 용이성, 코드 명확성 향상을 위해 대규모 리팩토링 진행

주요 변경 사항:

아키텍처 변경:

컨트롤러의 책임을 URL 매핑과 입력 검증으로 한정
서비스는 오로지 도메인 로직에만 집중(서비스와 Request DTO와의 결합 제거)
개별 유스케이스들을 총괄하는 ProductApplicationService 도입

컨트롤러는 이제 ProductApplicationService에만 의존하여 lean해짐

도메인 모델과 DTO 간 변환을 담당하는 ProductMapper를 빈(Bean)으로 전환하여 재사용성 향상

DTO 및 도메인 모델:

불변성과 간결성을 위해 ProductRequest, ProductResponse를 클래스에서 record 타입으로 마이그레이션

Product 모델 생성 로직을 캡슐화하기 위해 팩토리 메서드(of) 패턴 적용

테스트:

Deprecated로 사라질 예정인 MockBean을 제거.

통합 테스트(ProductTest)에 BDD(Given-When-Then) 스타일과 AAA 패턴 적용하여 테스트 목적과 시나리오 가독성 향상

Hyeonjun0527 committed:
refactor(product): 헥사고날 아키텍처 적용 및 기존 모듈 제거

- 헥사고날 아키텍처
spring-gift-wishlist/
└── src/
    └── main/
        └── java/
            └── gift/
                ├── common/                (공통 코드: 예외, 어노테이션, 유틸 등)
                └── product/
                    ├── adapter/           (인프라 계층: 외부와 연결)
                    │   ├── web/           (Primary Adapter: Controller 등)
                    │   └── persistence/   (Secondary Adapter: Repository 구현 등)
                    │
                    ├── application/       (애플리케이션 계층: 비즈니스 워크플로우)
                    │   ├── port/          (포트: 계층 간 추상화 인터페이스)
                    │   │   ├── in/        (Input Port: Use Case 인터페이스)
                    │   │   └── out/       (Output Port: 외부 의존 인터페이스)
                    │   └── usecase/       (Use Case 구현체: 실제 워크플로우/조정)
                    │
                    └── domain/            (도메인 계층: 핵심 비즈니스 로직, 엔티티, 도메인 서비스)

- 기존 애플리케이션 서비스 계층 삭제
- 포트와 어댑터 기반 아키텍처로 리팩터링: AddProductUseCase, DeleteProductUseCase 등 분리
- @Adapter와 UseCase 어노테이션 추가로 어댑터 역할 명시
- SOLID의 ISP 원칙에 따라 컨트롤러가 각 유스케이스 인터페이스를 개별 주입받도록 변경

Hyeonjun0527 committed:
feat(MemberPersistenceAdapter): Member Secondary Port 계층 구현
ProductPersistenceAdapter 계층 리팩토링

Hyeonjun0527 committed:
feat(authentication): JWT 기반 인증 및 회원가입/로그인 기능 추가

- JwtTokenPort 및 JwtTokenProvider 구현으로 JWT 생성, 검증, 정보 추출 기능 지원
- SimplePasswordEncoder
- LoginMemberUseCase, SignUpMemberUseCase 인터페이스
- MemberInteractor에서 회원가입 및 로그인 로직 구현
  - Access Token 및 Refresh Token 발급 기능
  - Refresh Token을 이용한 Access Token 갱신
- 회원 로그인 및 회원가입 요청/응답 DTO (LoginRequest, SignUpRequest, AuthResponse)
- 이전커밋에서 누락된 Member

Hyeonjun0527 committed:
refactor(member): SignUp을 Register로 이름 변경

Hyeonjun0527 committed:
refactor(jwt): 테스트 환경에서 MockJwtTokenProvider 추가 및 JWT 관련 설정 개선

- MockJwtTokenProvider 구현 및 테스트 프로파일 추가
  - 테스트 환경에서 JWT 토큰 생성을 위한 Mock 객체
- JwtTokenProvider 및 관련 빈들에 ConditionalOnProperty 및 Profile 설정
- WebConfig 테스트 프로파일
- Member domain에 Role 추가 및 관련 메서드 수정
  - Member 엔티티 및 관련 repository 메서드 리팩터링
  - schema.sql에 회원 테이블에 role 컬럼 추가 (기본값 USER)
- 테스트 환경 설정을 위해 각종 테스트 클래스에 ActiveProfiles("test") 추가
- JwtTokenPort 인터페이스 메서드 정리
- application-test.properties 추가 및 테스트 환경에서 jwt.enabled 설정 끔

Hyeonjun0527 committed:
fix(persistence):
- ProductPersistenceAdapter와 MemberPersistenceAdapter에서 기존 getKey() 메서드에서 getKeys 메서드에서 () 명시적으로 ID 키를 가져오도록 수정.
Member에 createdAt 때문에 keyHolder에 두가지 Key가 들어가기 때문
이메일 검증시 잘못된 이메일도 받았음. @Valid 어노테이션 빼먹어서 추가

Hyeonjun0527 committed:
refactor(core): 유스케이스 인터페이스 각각 MemberUseCase와 ProductUseCase로 통합
너무 생산성이 떨어지고 불편한 거 같아 제거 단일책임원칙을 지키는 것보다 내려놓는 게 좋다고 생각함

Hyeonjun0527 committed:
refactor(all):리뷰 반영 - 테스트 환경 bean 분리
  - TestConfig, MockJwtTokenProvider
  MockJwtTokenProvider는 MemberInteractor에서 사용중이기에 Mock가 필요했음.
  - admin.enabled, jwt.enabled 설정 값 조건부 로딩 추가
  - 현재 상황에 불필요한 @Order 제거
- ProductTest, ProductControllerTest에서 @ActiveProfiles 제거
- JwtTokenProvider에서 isValid 메서드를 isNotValid로 변경
- JwtAuthenticationFilter 가독성 향상
- WebConfig와 AuthorizationAspect 등에서 혼란을 야기하는 profile 기반 설정 +  ConditionalOnProperty에서 오직 ConditionalOnProperty로만 관리

Hyeonjun0527 committed:
feat(wish): 위시리스트 기능 추가

- 위시리스트 도메인, 테이블 정의
- 위시리스트 관련 포트 및 어댑터 추가
  - WishPersistencePort 인터페이스 추가
  - WishPersistenceAdapter 구현 추가
  - Product.of 정적 팩토리 메서드 도입
- README 업데이트
  - 위시리스트 기능 설명 및 구현 가이드 추가

Hyeonjun0527 committed:
feat(member): 회원 정보 업데이트 및 로그인 멤버 주입 기능 추가

- MemberMapper의 `toEntity` 메서드에서 null 체크 방식 수정
  - 기존: 삼항 연산자를 사용하여 null 검사
  - 변경: null 여부를 먼저 확인한 뒤 기존 값 반환
- `LoginMember` 어노테이션 추가
  - 컨트롤러 메서드의 파라미터에서 사용 가능하도록 설정
- `LoginMemberArgumentResolver` 구현
  - 로그인 사용자 정보를 주입하도록 ArgumentResolver 생성
  - `MemberUseCase`를 통해 로그인 사용자 로직 처리 준비
- 해당 기능 확장을 위한 클래스 및 어노테이션 추가

Hyeonjun0527 committed:
feat(member,wish): MemberMapper 정적 메서드화
- MemberMapper 메서드들을 정적으로 변경
- MemberInteractor에서 MemberMapper 의존성 제거
- MemberUseCase에 getMemberById 메서드 추가
- WebConfig에 LoginMemberArgumentResolver 등록
- LoginMemberArgumentResolver에서 request 객체 Null 체크 추가

Hyeonjun0527 committed:
feat(wish): 위시리스트 관리 기능 추가

- WishUseCase 인터페이스 추가
- Wish 도메인 모델 수정
  - Member, Product 연관관계 반영 및 수량
- WishMapper 추가: Wish 도메인을 WishResponse로 변환
- WishResponse DTO 추가: 위시리스트 반환 데이터 정의
- WishPersistencePort 추가
- schema.sql 수정: wish 테이블에 quantity 필드 추가
- MemberUseCase 및 MemberInteractor 수정: 이메일로 회원 정보를 가져오는 메서드 추가

Hyeonjun0527 committed:
feat(wish): 위시리스트 관리 기능 추가

- WishPersistenceAdapter 추가
- WishInteractor 구현
  - 회원별 위시리스트 목록 조회, 위시리스트 추가, 수량 변경 및 삭제
- Wish 관련 도메인 및 DTO 연관 매핑
  - Member, Product를 기반으로 Wish 생성 및 검증 로직 추가
- MemberInteractor 수정
  - 회원 조회 실패 시 명확한 예외 메시지 추가

Hyeonjun0527 committed:
feat(wish): 위시 관리 기능 추가 및 인증 처리 개선

- 위시 추가, 조회, 수량 변경, 삭제 기능 구현
  - WishController 추가
  - WishAddRequest, WishUpdateQuantityRequest DTO
- /login, /admin 페이지 처리 및 인증 로직 개선
  - PageController 추가
  - JwtAuthenticationFilter에서 인증 실패 시 /login 페이지 리다이렉트 처리
  - SKIP_PATHS에 /, /login 추가
- 관리자 페이지 접근 권한 관리 강화
  - AuthorizationAspect에서 /admin 페이지 접근권한 미충족 시 404 예외 처리
- Wish 관련 HTTP 요청 테스트 작성 (wish-api.http)
- 기본 데이터 추가: 관리자 계정, sample 상품 데이터 (`data.sql`)
- 기타 수정:
  - ProductRequest에서 imageUrl 검증 제거
  - LoginMemberArgumentResolver가 Member 대신 MemberResponse 반환하도록 수정
  - HTTP API 요청에 Authorization 헤더 추가

Hyeonjun0527 committed:
feat(admin): 관리자 페이지 UI 및 기능 추가

- 각종 네비게이션 문제 해결
    - 관리자페이지로 이동불가 문제 해결
    - url 접근은 허용하되 데이터는 허용하지 않도록 데이터 요청 전에 window.location.href=/login으로 보냄

- UI 구현, 메인페이지 / 구현 로그인 페이지 /login 구현, 관리자 페이지 /admin 구현

jaenyeong commented on src/main/java/gift/common/jwt/JwtAuthenticationFilter.java:
```diff
@@ -62,16 +67,25 @@ protected void doFilterInternal(HttpServletRequest request,
             if (validationResult.isNotValid()) {
                 log.warn("유효하지 않은 JWT 토큰: {}", validationResult.getErrorMessage());
-                sendUnauthorizedResponse(response, validationResult.getErrorMessage());
+                sendUnauthorizedResponse(request, response, validationResult.getErrorMessage());
```
🟢 
`JwtAuthenticationFilter` 이름 처럼 토큰 인증에만 집중하는 필터라고 가정한다면
여기서 JWT 토큰이 없거나 유효하지 않다면 `Authentication` 관련 네이밍이어야 할 것 같아요!
`Authorization` 은 결국 권한이 없어 인가/승인 받지 못한다는 뜻이기 때문에 이 경우는 네이밍이 조금 수정되는게 맞을 듯 하네요 ㅎㅎ

jaenyeong commented on src/main/java/gift/common/jwt/JwtAuthenticationFilter.java:
```diff
@@ -110,6 +131,13 @@ private void sendUnauthorizedResponse(HttpServletResponse response, String messa
+                return requestURI.startsWith(p);
+            }
+            return p.equals(requestURI);
+        });
```
🟢 
검증은 안해봤지만 ㅎㅎ...
```java
private final AntPathMatcher pathMatcher = new AntPathMatcher();

private static final List<String> PUBLIC_PATTERNS = Arrays.asList(
        "/",
        "/login",
        "/health",
        "/actuator/**",
        "/css/**",
        "/js/**",
        "/images/**",
        "/favicon.ico",
        "/api/members/register",
        "/api/members/login",
        "/h2-console/**"
);

private boolean isPublicUri(String requestURI) {
    return PUBLIC_PATTERNS.stream()
            .anyMatch(pattern -> pathMatcher.match(pattern, requestURI));
}
```
이런 식으로도 해볼 수 있지 않을까 싶네요? ㅎㅎ

jaenyeong commented on src/main/java/gift/member/adapter/web/mapper/MemberMapper.java:
```diff
@@ -9,21 +9,21 @@
     }
 
-    public Member toEntity(Member existingMember, UpdateMemberRequest request) {
+    public static Member toEntity(Member existingMember, UpdateMemberRequest request) {
```
❓ 
지금 코드에서는 두 메서드를 사용하지 않는 것 같은데 맞을까요?
그리고 스태틱 메서드로 활용하실 거라면 클래스도 `@Component` 필요 없을 것 같긴 하네요~

jaenyeong commented on src/main/java/gift/wish/adapter/persistence/persistenceAdapter/WishPersistenceAdapter.java:
```diff
@@ -0,0 +1,4 @@
+package gift.wish.adapter.persistence.persistenceAdapter;
+
+public class WishPersistenceAdapter {
+}
```
🔴 
이건 실수이실까요? ㅎㅎ 삭제가 필요한 것 같네요~

jaenyeong commented on src/main/java/gift/wish/adapter/persistence/WishPersistenceAdapter.java:
```diff
@@ -0,0 +1,142 @@
+                    product,
+                    rs.getInt("quantity")
+            );
+        };
```
🔴 
이러면 성능적으로 큰 문제가 있습니다 ㅎㅎ...
현준님이라면 이렇게만 말씀 드려도 이유가 왜인지 아시겠죠? :)

jaenyeong commented on src/main/java/gift/wish/application/port/in/dto/WishAddRequest.java:
```diff
@@ -0,0 +1,13 @@
+
+public record WishAddRequest(
+        @NotNull
+        Long productId,
```
🟢 
이것도 `quantity` 처럼 최솟값을 지정해줘도 좋을 것 같네요 ㅎㅎ

jaenyeong commented on src/main/java/gift/wish/adapter/persistence/WishPersistenceAdapter.java:
```diff
@@ -0,0 +1,142 @@
+        int start = pageable.getOffset();
+        if (start > totalRow) {
+            return new PageImpl<>(Collections.emptyList(), pageable, totalRow);
+        }
```
❓ 
```java
if (offset >= totalRows) {
    return new PageImpl<>(Collections.emptyList(), pageable, totalRows);
}
```
조건이 이렇게 되어야 하지 않을까요?
예를 들어 `totalRows`가 10이고 `size`가 5인 경우, 3페이지(offset=10)부터는 빈 결과여야 될 것 같은데
- 0페이지: offset = 0, 0~4 데이터 표시
- 1페이지: offset = 5, 5~9 데이터 표시
- 2페이지: offset = 10 -> 표시할 데이터가 없어 빈 페이지 노출
제가 잘못 이해했다면 말씀 부탁드려요 ㅎㅎ

그리고 `start` 변수명보다는 `offset` 이라는 이름을 쓰셔서 쿼리의 offset을 바로 의미해도 좋을 듯 하네용~

jaenyeong commented on src/main/java/gift/common/config/LoginMemberArgumentResolver.java:
```diff
@@ -0,0 +1,45 @@
+            throw new IllegalArgumentException("인증된 사용자 정보를 찾을 수 없습니다.");
+        }
+
+        return memberUseCase.getMemberById(memberId);
```
❓ 
멤버의 아이디로 찾는데 이메일은 없어도 되지 않을까요?? 이메일 체크 로직을 넣으신 의도가 있으실까요?

jaenyeong commented on src/main/java/gift/wish/application/usecase/WishInteractor.java:
```diff
@@ -0,0 +1,76 @@
+        Wish wish = wishPersistencePort.findById(wishId)
+                .orElseThrow(() -> new IllegalArgumentException("Wish not found"));
+        if (!wish.getMember().id().equals(memberId)) {
+            throw new SecurityException("Not authorized to update this wish");
```
🟢 
저는 개인적으로 여기에서 시큐리티 익셉션을 던지는 것 보다 직접 구현하신 `UnauthorizedException` 를 활용하는 게 더 직관적일 것 같다는 느낌이 드네요 ㅎㅎ
해당 예외도 아마 권한 등을 문제로 던지는 예외긴 하겠지만, 파일 시스템에 직접 접근한다든지 보안과 관련된 경우를 의미할 수 있다보니 조금 더 직관적인 예외로 핸들링하는게 좋지 않을까 생각 합니다 ㅎㅎ

jaenyeong commented on src/main/java/gift/wish/application/usecase/WishInteractor.java:
```diff
@@ -0,0 +1,76 @@
+                .orElseThrow(() -> new IllegalArgumentException("Wish not found"));
+        if (!wish.getMember().id().equals(memberId)) {
+            throw new SecurityException("Not authorized to delete this wish");
+        }
```
🟢 
멤버만 본인인지 확인하고 삭제처리 해서 멱등성을 지켜줘도 큰 문제는 없을 것 같네요 ㅎㅎ

jaenyeong reviewed (changes requested):
리뷰가 조금 늦었습니다 ㅠㅠ 잘해주셨는데 🔴 코멘트 수정 부탁드리겠습니다!

질문에 답변을 조금 드리자면 로컬스토리지에 있는 값 자체를 수정할 수 있기 때문에 보안적으로 좋다고 보기는 어렵겠습니다 ㅎㅎㅎ 그래서 말씀하신대로 쿠키를 활용하는 것도 방법 중에 하나가 될 수 있겠네요!

그리고 상품 조회 정도는 로그인하지 않아도 조회가 가능하게 두어도 되지 않을까 싶네요 ㅎㅎ

Hyeonjun0527 committed:
refactor(auth, wish): 개선된 예외 처리 및 SQL 쿼리 리팩토링

- UnauthorizedException 로 에러메세지 변경
- WishPersistenceAdapter 내부에 RowMapper 리팩토링
  - WISH_WITH_DETAILS_ROW_MAPPER 추가 중복 로직 공통화
- JwtAuthenticationFilter의
  - sendUnauthorizedResponse → sendUnAuthenticationResponse로 이름 변경
  - PUBLIC_PATTERNS로 public endpoint 패턴 등록 및 AntPathMatcher로 필터 처리 일관성 강화
- @Min, @NotNull 어노테이션 추가
- Member/Adapter/Persistence 경로 구조 리팩토링
  - persistenceAdapter → adapter로 경로 변경

Hyeonjun0527 committed:
refactor(auth, wish): 개선된 예외 처리 및 SQL 쿼리 리팩토링

- UnauthorizedException 로 에러메세지 변경
- WishPersistenceAdapter 내부에 RowMapper 리팩토링
  - WISH_WITH_DETAILS_ROW_MAPPER 추가 중복 로직 공통화
- JwtAuthenticationFilter의
  - sendUnauthorizedResponse → sendUnAuthenticationResponse로 이름 변경
  - PUBLIC_PATTERNS로 public endpoint 패턴 등록 및 AntPathMatcher로 필터 처리 일관성 강화
- @Min, @NotNull 어노테이션 추가
- Member/Adapter/Persistence 경로 구조 리팩토링
  - persistenceAdapter → adapter로 경로 변경

jaenyeong reviewed (approved):
마지막 스텝까지 고생 많으셨습니다 현준님~
사실 헥사고날을 적용하면서 자체 난이도를 높혀서 다른 수강생 분들과 다르게 진행하시게 되었는데 소감이 어떠신지도 궁금하네요 ㅎㅎㅎ 이 부분은 제가 교육 커리큘럼에 벗어나기도 하고, 이것을 피드백을 하게되면 다른 기초 부분들 쪽으로 피드백을 집중해드릴 수가 없어서 아쉽게도 많이 코멘트 드리진 못했습니다 ㅎㅎ 이렇게 복잡한 형태로 계속 학습해 보시다 보면 간결하고 직관적인게 최고구나라고 돌아오실 때가 있으실 것 같긴 합니다만.. ㅎㅎ

그래도 공부하고 학습하실 때 다양한 시도를 해보시는 것은 제 개인적으로는 좋은 학습 방법이라고 생각해서 현준님을 응원하게 되네요~

마지막 피드백도 잘 반영해 주셔서 감사드리고 조금 진도가 늦어졌지만 다음 미션도 화이팅 하셨으면 좋겠습니다!
수고 많으셨어요~
